File: packages/sdk/sdk-go/parameter_support.go
package phaseo

import (
	"context"
	"encoding/json"
	"fmt"
	"math"
	"strings"

	gen "github.com/phaseoteam/Phaseo/packages/sdk/sdk-go/v3/src/gen"
)

type ParameterSupportStatus string

const (
	ParameterSupported   ParameterSupportStatus = "supported"
	ParameterPartial     ParameterSupportStatus = "partial"
	ParameterUnsupported ParameterSupportStatus = "unsupported"
	ParameterUnknown     ParameterSupportStatus = "unknown"
)

type ParameterRouteReference struct {
	ID         string `json:"id"`
	Provider   string `json:"provider"`
	Endpoint   string `json:"endpoint"`
	PublicPath string `json:"public_path"`
}

type ParameterConstraint struct {
	Route  ParameterRouteReference `json:"route"`
	Detail map[string]any          `json:"detail"`
}

type ParameterSupport struct {
	Name          string                    `json:"name"`
	Value         any                       `json:"value"`
	Status        ParameterSupportStatus    `json:"status"`
	SupportedBy   []ParameterRouteReference `json:"supported_by"`
	AcceptedBy    []ParameterRouteReference `json:"accepted_by"`
	UnsupportedBy []ParameterRouteReference `json:"unsupported_by"`
	Constraints   []ParameterConstraint     `json:"constraints"`
	Issues        []string                  `json:"issues"`
}

type ParameterSupportOptions struct {
	Endpoint  string
	Providers []string
}

type ParameterSupportReport struct {
	OK             bool                      `json:"ok"`
	ModelID        string                    `json:"model_id"`
	RouteCount     int                       `json:"route_count"`
	MatchingRoutes []ParameterRouteReference `json:"matching_routes"`
	Parameters     []ParameterSupport        `json:"parameters"`
	Issues         []string                  `json:"issues"`
}

type PreflightReport struct {
	OK                bool                   `json:"ok"`
	ModelID           string                 `json:"model_id"`
	CheckedParameters map[string]any         `json:"checked_parameters"`
	ParameterSupport  ParameterSupportReport `json:"parameter_support"`
}

var preflightStructuralFields = map[string]bool{
	"model": true, "input": true, "messages": true, "prompt": true, "contents": true,
	"provider": true, "providers": true, "routing": true, "metadata": true,
	"session_id": true, "app": true, "webhook": true, "idempotency_key": true,
}

func splitModelID(modelID string) (map[string]string, error) {
	parts := strings.SplitN(strings.TrimSpace(modelID), "/", 2)
	if len(parts) != 2 || parts[0] == "" || parts[1] == "" {
		return nil, fmt.Errorf("model ID must use author/slug format")
	}
	return map[string]string{"author": parts[0], "slug": parts[1]}, nil
}

// GetModelEndpointCapabilities returns live, provider-specific endpoint metadata for a model.
func (c *Phaseo) GetModelEndpointCapabilities(_ context.Context, modelID string, query map[string]string) (map[string]interface{}, error) {
	path, err := splitModelID(modelID)
	if err != nil {
		return nil, err
	}
	return gen.ListModelEndpoints(c.raw, path, query, nil, nil)
}

// CheckModelParameters reports which active routes support and accept the supplied values.
func (c *Phaseo) CheckModelParameters(ctx context.Context, modelID string, values map[string]any, options ParameterSupportOptions) (ParameterSupportReport, error) {
	payload, err := c.GetModelEndpointCapabilities(ctx, modelID, nil)
	if err != nil {
		return ParameterSupportReport{}, err
	}
	return CheckParameterSupport(payload, values, options), nil
}

// PreflightRequest checks every non-structural request field against live model routes.
func (c *Phaseo) PreflightRequest(ctx context.Context, request map[string]any, options ParameterSupportOptions) (PreflightReport, error) {
	modelID := strings.TrimSpace(asString(request["model"]))
	if modelID == "" { return PreflightReport{}, fmt.Errorf("preflight requires request model") }
	values := map[string]any{}
	for name, value := range request {
		if !preflightStructuralFields[name] && value != nil { values[name] = value }
	}
	support, err := c.CheckModelParameters(ctx, modelID, values, options)
	if err != nil { return PreflightReport{}, err }
	return PreflightReport{OK: support.OK, ModelID: modelID, CheckedParameters: values, ParameterSupport: support}, nil
}

// CheckParameterSupport builds a UI-friendly report from a model endpoint response.
func CheckParameterSupport(model map[string]any, values map[string]any, options ParameterSupportOptions) ParameterSupportReport {
	providerFilter := map[string]bool{}
	for _, provider := range options.Providers {
		providerFilter[provider] = true
	}
	routes := []map[string]any{}
	for _, raw := range asSlice(model["endpoints"]) {
		route, ok := raw.(map[string]any)
		if !ok {
			continue
		}
		if route["routable"] != true || asString(route["status"]) != "active" {
			continue
		}
		provider := asString(asMap(route["provider"])["id"])
		if len(providerFilter) > 0 && !providerFilter[provider] {
			continue
		}
		if options.Endpoint != "" && !endpointMatches(route, options.Endpoint) {
			continue
		}
		routes = append(routes, route)
	}
	parameters := make([]ParameterSupport, 0, len(values))
	for name, value := range values {
		item := ParameterSupport{Name: name, Value: value, SupportedBy: []ParameterRouteReference{}, AcceptedBy: []ParameterRouteReference{}, UnsupportedBy: []ParameterRouteReference{}, Constraints: []ParameterConstraint{}, Issues: []string{}}
		knownRoutes := 0
		valueIssues := []string{}
		for _, route := range routes {
			ref := routeReference(route)
			capabilities := asMap(route["capabilities"])
			advertised, known := stringSlice(capabilities["parameters"])
			if !known {
				item.UnsupportedBy = append(item.UnsupportedBy, ref)
				continue
			}
			knownRoutes++
			detail := asMap(asMap(capabilities["parameter_details"])[name])
			if !contains(advertised, name) && detail["supported"] != true || detail["supported"] == false {
				item.UnsupportedBy = append(item.UnsupportedBy, ref)
				continue
			}
			item.SupportedBy = append(item.SupportedBy, ref)
			if len(detail) > 0 {
				item.Constraints = append(item.Constraints, ParameterConstraint{Route: ref, Detail: detail})
			}
			issues := parameterValueIssues(name, value, detail)
			if len(issues) == 0 {
				item.AcceptedBy = append(item.AcceptedBy, ref)
			} else {
				valueIssues = append(valueIssues, issues...)
			}
		}
		switch {
		case len(routes) == 0 || knownRoutes == 0:
			item.Status = ParameterUnknown
		case len(item.SupportedBy) == 0:
			item.Status = ParameterUnsupported
		case len(item.SupportedBy) == len(routes):
			item.Status = ParameterSupported
		default:
			item.Status = ParameterPartial
		}
		if item.Status == ParameterUnsupported {
			item.Issues = []string{fmt.Sprintf("%s is not supported by any matching active route", name)}
		} else if len(item.SupportedBy) > 0 && len(item.AcceptedBy) == 0 {
			item.Issues = unique(valueIssues)
		}
		parameters = append(parameters, item)
	}
	matching := []ParameterRouteReference{}
	for _, route := range routes {
		ref := routeReference(route)
		if everyAccepted(parameters, ref.ID) {
			matching = append(matching, ref)
		}
	}
	issues := []string{}
	if len(routes) == 0 {
		issues = append(issues, "No active routable model endpoints matched the filters")
	} else if len(matching) == 0 && len(parameters) > 0 {
		issues = append(issues, "No active route supports all requested parameter values together")
	}
	for _, parameter := range parameters {
		issues = append(issues, parameter.Issues...)
	}
	modelID := asString(model["id"])
	if modelID == "" {
		modelID = "unknown"
	}
	return ParameterSupportReport{OK: len(routes) > 0 && len(matching) > 0, ModelID: modelID, RouteCount: len(routes), MatchingRoutes: matching, Parameters: parameters, Issues: unique(issues)}
}

func routeReference(route map[string]any) ParameterRouteReference {
	provider := asString(asMap(route["provider"])["id"])
	if provider == "" {
		provider = "unknown"
	}
	endpoint := asString(route["endpoint"])
	if endpoint == "" {
		endpoint = asString(route["capability_id"])
	}
	id := asString(route["id"])
	if id == "" {
		id = provider + ":" + endpoint
	}
	publicPath := asString(route["public_path"])
	if publicPath == "" {
		publicPath = endpoint
	}
	return ParameterRouteReference{ID: id, Provider: provider, Endpoint: endpoint, PublicPath: publicPath}
}
func endpointMatches(route map[string]any, requested string) bool {
	normalized := strings.TrimPrefix(strings.TrimPrefix(requested, "/v1/"), "/")
	for _, value := range []string{asString(route["endpoint"]), asString(route["capability_id"]), strings.TrimPrefix(strings.TrimPrefix(asString(route["public_path"]), "/v1/"), "/")} {
		if value == requested || value == normalized {
			return true
		}
	}
	return false
}
func parameterValueIssues(name string, value any, detail map[string]any) []string {
	issues := []string{}
	if detail["supported"] == false {
		issues = append(issues, name+" is unsupported")
	}
	allowed := asSlice(detail["values"])
	if len(allowed) == 0 {
		allowed = asSlice(detail["enum"])
	}
	if len(allowed) > 0 {
		encoded, _ := json.Marshal(value)
		found := false
		for _, item := range allowed {
			other, _ := json.Marshal(item)
			if string(encoded) == string(other) {
				found = true
				break
			}
		}
		if !found {
			parts := []string{}
			for _, item := range allowed {
				parts = append(parts, fmt.Sprint(item))
			}
			issues = append(issues, fmt.Sprintf("%s must be one of: %s", name, strings.Join(parts, ", ")))
		}
	}
	number, ok := asFloat(value)
	if ok {
		if math.IsInf(number, 0) || math.IsNaN(number) {
			issues = append(issues, name+" must be finite")
		}
		if minimum, ok := asFloat(detail["minimum"]); ok && number < minimum {
			issues = append(issues, fmt.Sprintf("%s must be at least %v", name, minimum))
		}
		if maximum, ok := asFloat(detail["maximum"]); ok && number > maximum {
			issues = append(issues, fmt.Sprintf("%s must be at most %v", name, maximum))
		}
		if step, ok := asFloat(detail["step"]); ok && step > 0 {
			minimum, _ := asFloat(detail["minimum"])
			steps := (number - minimum) / step
			if math.Abs(steps-math.Round(steps)) > 1e-8 {
				issues = append(issues, fmt.Sprintf("%s must use steps of %v", name, step))
			}
		}
	}
	return issues
}
func asFloat(v any) (float64, bool) {
	switch n := v.(type) {
	case float64:
		return n, true
	case float32:
		return float64(n), true
	case int:
		return float64(n), true
	case int64:
		return float64(n), true
	default:
		return 0, false
	}
}
func stringSlice(v any) ([]string, bool) {
	raw, ok := v.([]any)
	if !ok {
		return nil, false
	}
	out := make([]string, 0, len(raw))
	for _, item := range raw {
		if text, ok := item.(string); ok {
			out = append(out, text)
		}
	}
	return out, true
}
func contains(values []string, target string) bool {
	for _, value := range values {
		if value == target {
			return true
		}
	}
	return false
}
func everyAccepted(parameters []ParameterSupport, routeID string) bool {
	for _, parameter := range parameters {
		found := false
		for _, route := range parameter.AcceptedBy {
			if route.ID == routeID {
				found = true
				break
			}
		}
		if !found {
			return false
		}
	}
	return true
}
func unique(values []string) []string {
	seen := map[string]bool{}
	out := []string{}
	for _, value := range values {
		if !seen[value] {
			seen[value] = true
			out = append(out, value)
		}
	}
	return out
}

File: packages/sdk/sdk-rust/src/parameter_support.rs
use std::collections::{HashMap, HashSet};

use serde_json::{json, Value};

#[derive(Clone, Debug, Default, PartialEq)]
pub struct ParameterSupportOptions {
    pub endpoint: Option<String>,
    pub providers: Vec<String>,
}

/// Builds a UI-friendly report from live model endpoint metadata.
pub fn check_parameter_support(
    model: &Value,
    values: &HashMap<String, Value>,
    options: &ParameterSupportOptions,
) -> Value {
    let providers: HashSet<&str> = options.providers.iter().map(String::as_str).collect();
    let requested_endpoint = options.endpoint.as_deref().map(normalize_endpoint);
    let routes: Vec<&Value> = model["endpoints"]
        .as_array()
        .into_iter()
        .flatten()
        .filter(|route| {
            route["routable"].as_bool() == Some(true)
                && route["status"].as_str() == Some("active")
                && (providers.is_empty()
                    || providers.contains(route["provider"]["id"].as_str().unwrap_or("")))
                && requested_endpoint
                    .as_ref()
                    .map(|endpoint| endpoint_matches(route, endpoint))
                    .unwrap_or(true)
        })
        .collect();
    let mut parameters = Vec::new();
    for (name, value) in values {
        let mut supported_by = Vec::new();
        let mut accepted_by = Vec::new();
        let mut unsupported_by = Vec::new();
        let mut constraints = Vec::new();
        let mut value_issues = Vec::new();
        let mut known_routes = 0;
        for route in &routes {
            let reference = route_reference(route);
            let Some(advertised) = route["capabilities"]["parameters"].as_array() else {
                unsupported_by.push(reference);
                continue;
            };
            known_routes += 1;
            let detail = &route["capabilities"]["parameter_details"][name];
            let supports = advertised.iter().any(|item| item.as_str() == Some(name))
                || detail["supported"].as_bool() == Some(true);
            if !supports || detail["supported"].as_bool() == Some(false) {
                unsupported_by.push(reference);
                continue;
            }
            supported_by.push(reference.clone());
            if detail.as_object().is_some_and(|object| !object.is_empty()) {
                constraints.push(json!({"route": reference, "detail": detail}));
            }
            let issues = parameter_value_issues(name, value, detail);
            if issues.is_empty() {
                accepted_by.push(reference);
            } else {
                value_issues.extend(issues);
            }
        }
        let status = if routes.is_empty() || known_routes == 0 {
            "unknown"
        } else if supported_by.is_empty() {
            "unsupported"
        } else if supported_by.len() == routes.len() {
            "supported"
        } else {
            "partial"
        };
        let issues = if status == "unsupported" {
            vec![format!(
                "{name} is not supported by any matching active route"
            )]
        } else if !supported_by.is_empty() && accepted_by.is_empty() {
            unique(value_issues)
        } else {
            Vec::new()
        };
        parameters.push(json!({"name":name,"value":value,"status":status,"supported_by":supported_by,"accepted_by":accepted_by,"unsupported_by":unsupported_by,"constraints":constraints,"issues":issues}));
    }
    let matching_routes: Vec<Value> = routes
        .iter()
        .filter_map(|route| {
            let reference = route_reference(route);
            let id = reference["id"].as_str()?;
            parameters
                .iter()
                .all(|parameter| {
                    parameter["accepted_by"].as_array().is_some_and(|items| {
                        items.iter().any(|item| item["id"].as_str() == Some(id))
                    })
                })
                .then_some(reference)
        })
        .collect();
    let mut issues = Vec::new();
    if routes.is_empty() {
        issues.push("No active routable model endpoints matched the filters".to_string());
    } else if matching_routes.is_empty() && !parameters.is_empty() {
        issues.push("No active route supports all requested parameter values together".to_string());
    }
    for parameter in &parameters {
        for issue in parameter["issues"].as_array().into_iter().flatten() {
            if let Some(issue) = issue.as_str() {
                issues.push(issue.to_string())
            }
        }
    }
    json!({"ok":!routes.is_empty()&&!matching_routes.is_empty(),"model_id":model["id"].as_str().unwrap_or("unknown"),"route_count":routes.len(),"matching_routes":matching_routes,"parameters":parameters,"issues":unique(issues)})
}

fn route_reference(route: &Value) -> Value {
    let provider = route["provider"]["id"].as_str().unwrap_or("unknown");
    let endpoint = route["endpoint"]
        .as_str()
        .or_else(|| route["capability_id"].as_str())
        .unwrap_or("unknown");
    json!({"id":route["id"].as_str().map(str::to_owned).unwrap_or_else(||format!("{provider}:{endpoint}")),"provider":provider,"endpoint":endpoint,"public_path":route["public_path"].as_str().unwrap_or(endpoint)})
}
fn normalize_endpoint(value: &str) -> String {
    value
        .strip_prefix("/v1/")
        .or_else(|| value.strip_prefix('/'))
        .unwrap_or(value)
        .to_string()
}
fn endpoint_matches(route: &Value, endpoint: &str) -> bool {
    route["endpoint"].as_str() == Some(endpoint)
        || route["capability_id"].as_str() == Some(endpoint)
        || normalize_endpoint(route["public_path"].as_str().unwrap_or("")) == endpoint
}
fn parameter_value_issues(name: &str, value: &Value, detail: &Value) -> Vec<String> {
    let mut issues = Vec::new();
    let allowed = detail
        .get("values")
        .or_else(|| detail.get("enum"))
        .and_then(Value::as_array);
    if allowed.is_some_and(|items| !items.contains(value)) {
        issues.push(format!(
            "{name} must be one of: {}",
            allowed
                .unwrap()
                .iter()
                .map(Value::to_string)
                .collect::<Vec<_>>()
                .join(", ")
        ));
    }
    if let Some(number) = value.as_f64() {
        if !number.is_finite() {
            issues.push(format!("{name} must be finite"));
        }
        if let Some(min) = detail["minimum"].as_f64() {
            if number < min {
                issues.push(format!("{name} must be at least {min}"));
            }
        }
        if let Some(max) = detail["maximum"].as_f64() {
            if number > max {
                issues.push(format!("{name} must be at most {max}"));
            }
        }
        if let Some(step) = detail["step"].as_f64() {
            if step > 0.0 {
                let start = detail["minimum"].as_f64().unwrap_or(0.0);
                let steps = (number - start) / step;
                if (steps - steps.round()).abs() > 1e-8 {
                    issues.push(format!("{name} must use steps of {step}"));
                }
            }
        }
    }
    issues
}
fn unique(values: Vec<String>) -> Vec<String> {
    let mut seen = HashSet::new();
    values
        .into_iter()
        .filter(|value| seen.insert(value.clone()))
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn highlights_invalid_value() {
        let model = json!({"id":"openai/example","endpoints":[{"id":"openai:responses","endpoint":"responses","routable":true,"status":"active","provider":{"id":"openai"},"capabilities":{"parameters":["temperature"],"parameter_details":{"temperature":{"supported":true,"minimum":0,"maximum":1}}}}]});
        let values = HashMap::from([("temperature".to_string(), json!(1.5))]);
        let report = check_parameter_support(&model, &values, &ParameterSupportOptions::default());
        assert_eq!(report["ok"], false);
        assert_eq!(report["parameters"][0]["status"], "supported");
        assert_eq!(report["parameters"][0]["accepted_by"], json!([]));
    }
}
